### PR TITLE
Enhancement/update linter

### PR DESCRIPTION
**Pull request by tkah**
* **Problem:**
   1. ESLint version, and associated plugins, have fallen behind latest releases.
* **Solution:**
   1. Updated the following packages `eslint`, `eslint-config-airbnb`, `eslint-plugin-import`, `eslint-plugin-jsx-a11y`, `eslint-plugin-react` and modified codebase to match new rules.
      1. Plugins don't yet support newest version [v4] of `eslint`, meaning `eslint` is at `3.19.0`.
* **Tests:**
   1. `npm run lint` and `npm run test` in base directory
   1. `npm run test:e2e` in coinstac-ui package directory.

### CODE REVIEW

File: packages/coinstac-client-core/bin/coinstac-client-core [hunk 1]
```diff
@@ -168,7 +169,7 @@ program
 
     return Promise.all([
       getFiles(program.args).then(client.projects.getFileStats),
-      client.projects.getCSV(program.metafile),
+      client.projects.constructor.getCSV(program.metafile),
```
Comment (swashcap) [inline packages/coinstac-client-core/bin/coinstac-client-core:172]: _Gross_ but understandable.

File: packages/coinstac-client-core/package.json [hunk 2]
```diff
@@ -41,7 +41,9 @@
     "commander": "^2.9.0",
     "convict": "^1.5.0",
     "csv-parse": "^1.2.0",
+    "cycle": "^1.0.3",
```
Comment (swashcap) [inline packages/coinstac-client-core/package.json:44]: Why the new dependency?

File: packages/coinstac-client-core/src/model-service.js [hunk 3]
```diff
@@ -35,11 +36,13 @@ class ModelService {
     this.client = opts.client;
 
     /* istanbul ignore if */
-    if (!this.modelServiceHooks || !(this.modelServiceHooks instanceof Function)) {
+    if (!this.constructor.modelServiceHooks
```
Comment (swashcap) [inline packages/coinstac-client-core/src/model-service.js]: I typically break these up so that the boolean operator terminates the line and conditions are indented at equal levels. So:

```js
if (
  !this.constructor.modelServiceHooks ||
  !(this.constructor.modelServiceHooks instanceof Function)
) {
  // ...
}
```

As it turns out, two checks are redundant here. We don't need the lookup optimization before checking that `modelServiceHooks` is a function. So:

```js
if (!(this.constructor.modelServiceHooks instanceof Function)) {
  // ...
}
```

File: packages/coinstac-client-core/src/model-service.js [hunk 4]
```diff
@@ -81,7 +84,7 @@ class ModelService {
   /**
    * @abstract
    */
-  modelServiceHooks() {
+  static modelServiceHooks() {
```
Comment (swashcap) [inline packages/coinstac-client-core/src/model-service.js]: Bleh. This was super _gross_ to begin with, but making it `static` just makes it worse. Not sure it’s worth addressing in this PR.

File: packages/coinstac-client-core/test/e2e/client.js [hunk 5]
```diff
@@ -54,7 +54,7 @@ function getQueueEndStopper(count, fn) {
        *
        * @todo Determine if this is necessary
        */
-      if (typeof value === 'object' && value.then & value.catch) {
+      if (typeof value === 'object' && value.then && value.catch) {
```
Comment (swashcap) [inline packages/coinstac-client-core/test/e2e/client.js:57]: _Noice._

File: packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js [hunk 6]
```diff
@@ -407,13 +408,13 @@ class PipelineRunnerPool extends Base {
       ].join(' '));
     }
     /* istanbul ignore next */
-    if (typeof this.resultsListeners[db.name] === DBListener) {
+    if (typeof this.resultsListeners[db.name] === typeof DBListener) {
```
Comment (swashcap) [inline packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js]: Bad comparison. This should use the `instanceof` operator:

```js
if (this.resultsListeners[db.name] instanceof DBListener) {
  // ...
}
```

File: packages/coinstac-common/test/models/pipeline/runner/remote-pipeline-runner.js [hunk 7]
```diff
@@ -40,7 +41,7 @@ test('RemotePipelineRunner::run - basic', t => {
     const pipelineRunStub = () => {
       // sinon stub is freaking out, so self-stub
       pipelineRunStub.called = true;
-      runner.pipeline.run.apply(runner.pipeline, arguments);
+      runner.pipeline.run(...arguments);
```
Comment (swashcap) [inline packages/coinstac-common/test/models/pipeline/runner/remote-pipeline-runner.js:44]: Lambdas (arrow fns) lack lexical scope. `arguments` is going to point to something unexpected here.

File: packages/coinstac-common/test/models/pouch-document.js [hunk 8]
```diff
@@ -1,10 +1,9 @@
 'use strict';
 
-const path = require('path');
-const PD = require(path.join(__dirname, '../../src/models/pouch-document'));
```
Comment (swashcap) [inline packages/coinstac-common/test/models/pouch-document.js:4]: Wow, this was…strange. Nice fix.

File: packages/coinstac-server-core/test/executable.js [hunk 9]
```diff
@@ -10,23 +10,23 @@ const executable = path.join(__dirname, '..', 'bin', 'coinstac-server-core');
...
 }
 
-tape('executable works', t => {
+tape('executable works', (t) => {
   t.plan(2);
 
   try {
     const child = spawn(executable);
-    const errorExit = exitCode => { t.notOk(exitCode, 'exited with non-error'); };
+    const errorExit = (exitCode) => { t.notOk(exitCode, 'exited with non-error'); };
```
Comment (swashcap) [inline packages/coinstac-server-core/test/executable.js]: Surprised the linter doesn’t bark at this brace-expression organization. Want to break it up with some `\n`s to make it easier to read?

File: packages/coinstac-ui/app/render/components/consortium/consortium-computation-selector.jsx [hunk 10]
```diff
@@ -33,29 +33,28 @@ export default function ConsortiumComputationSelector({
...
-            index
+            }
           ) => {
             const isChecked = input.value === _id;
 
             return (
-              <div className={radioClassNames} key={index}>
-                <label>
+              <div className={radioClassNames} key={_id}>
+                <label htmlFor={`computation-radio-${_id}`}>
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/consortium/consortium-computation-selector.jsx:44]: Nice.

File: packages/coinstac-ui/app/render/components/consortium/consortium-computation-selector.jsx [hunk 11]
```diff
@@ -33,29 +33,28 @@ export default function ConsortiumComputationSelector({
           (
             {
               _id,
-              meta: { description, name, tags },
+              meta: { description, name }, // , tags },
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/consortium/consortium-computation-selector.jsx]: Want to just remove the commented out portion?

File: packages/coinstac-ui/app/render/components/consortium/consortium-form.jsx [hunk 12]
```diff
@@ -218,18 +218,18 @@ class ConsortiumForm extends Component {
 ConsortiumForm.displayName = 'ConsortiumForm';
 
 ConsortiumForm.propTypes = {
-  activeComputationId: PropTypes.string,
+  // activeComputationId: PropTypes.string,
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/consortium/consortium-form.jsx]: Do we not need these props any more? Just remove them entirely?

File: packages/coinstac-ui/app/render/components/consortium/consortium-result.jsx [hunk 13]
```diff
@@ -193,13 +193,18 @@ ConsortiumResult.propTypes = {
   }),
   endDate: PropTypes.number,
   expanded: PropTypes.bool.isRequired,
-  pipelineState: PropTypes.object.isRequired,
+  // pipelineState: PropTypes.object.isRequired,
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/consortium/consortium-result.jsx]: Do we not reference `pipelineState`? Just drop it?

File: packages/coinstac-ui/app/render/components/dashboard-consortia.jsx [hunk 14]
```diff
@@ -2,8 +2,8 @@ import React, { Component } from 'react';
 import PropTypes from 'prop-types';
 import { Button } from 'react-bootstrap';
 import { LinkContainer } from 'react-router-bootstrap';
-import ConsortiumCard from './consortium-card';
 import { connect } from 'react-redux';
+import ConsortiumCard from './consortium-card';
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/dashboard-consortia.jsx:6]: So particular about import order. 🔍😵

File: packages/coinstac-ui/app/render/components/dashboard.jsx [hunk 15]
```diff
@@ -36,7 +36,7 @@ class Dashboard extends Component {
       <div className="dashboard container-fluid">
         <div className="row">
           <div className="col-xs-12 col-sm-4">
-            <nav className="navigation" role="navigation">
+            <nav className="navigation">
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/dashboard.jsx:39]: aria `role`s not supported?

File: packages/coinstac-ui/app/render/components/field-password.jsx [hunk 16]
```diff
@@ -20,8 +20,8 @@ export default class FieldPassword extends Component {
   }
 
   handleChange() {
-    const password = findDOMNode(this.refs.password).value;
-    const confirmPassword = findDOMNode(this.refs.confirm).value;
+    const password = findDOMNode(this.formPassword).value;
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/field-password.jsx:23]: Why this change?

File: packages/coinstac-ui/app/render/components/projects/form-project.jsx [hunk 17]
```diff
@@ -339,7 +339,7 @@ export default class FormProject extends Component {
 }
 
 FormProject.propTypes = {
-  allowComputationRun: PropTypes.bool.isRequired,
+  // allowComputationRun: PropTypes.bool.isRequired,
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/projects/form-project.jsx]: Not using these properties?

File: packages/coinstac-ui/app/render/components/projects/project-covariates-mapper.jsx [hunk 18]
```diff
@@ -130,3 +129,9 @@ ProjectCovariatesMapper.propTypes = {
   metaCovariateMapping: PropTypes.object.isRequired,
   onMapCovariate: PropTypes.func.isRequired,
 };
+
+ProjectCovariatesMapper.defaultProps = {
+  metaCovariateErrors: null,
+};
+
+export default ProjectCovariatesMapper;
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/projects/project-covariates-mapper.jsx]: Why move the `export` down here?

File: packages/coinstac-ui/app/render/index.js [hunk 19]
```diff
@@ -28,10 +27,9 @@ app.analysisRequestId = app.analysisRequestId || 0;
 require('./styles/app.scss');
 
 const rootEl = document.getElementById('app');
-const store = configureStore();
 
 render(
-  <Root history={hashHistory} store={store} />,
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/index.js]: Why this change? Seems like this will break hot reloading as the store and history are created anew in `Root.jsx` versus here, where they were created in a higher closure for reuse for Webpack’s hot reloading machinery.

File: packages/coinstac-ui/app/render/state/ducks/util.js [hunk 20]
```diff
@@ -1,10 +1,12 @@
 import app from 'ampersand-app';
 
-export const notifyAndThrow = (err) => {
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/state/ducks/util.js:3]: Is `export const <expression>` no longer allowed? 😕

**Review by swashcap — CHANGES_REQUESTED** [review]
💖 Nice work! It’s a bit of a slog, but it’s lookin’ good.

File: packages/coinstac-client-core/package.json (hunk 2 above)
Comment (tkah) [reply packages/coinstac-client-core/package.json:44]: It was reported by the linter for not being included in the `package.json`, but being imported in one of the files (`dynamic-api.js`).

File: packages/coinstac-client-core/src/model-service.js (hunk 4 above)
Comment (tkah) [reply packages/coinstac-client-core/src/model-service.js]: You want me to pull it out then? It was one of the things reported by the linter: http://eslint.org/docs/rules/class-methods-use-this

File: packages/coinstac-common/test/models/pipeline/runner/remote-pipeline-runner.js (hunk 7 above)
Comment (swashcap) [reply packages/coinstac-common/test/models/pipeline/runner/remote-pipeline-runner.js:44]: I’ll look into this. Is this stub even necessary?

File: packages/coinstac-ui/app/render/components/dashboard.jsx (hunk 15 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/dashboard.jsx:39]: They are. The issue is including a role of `navigation` on a `nav` element: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-redundant-roles.md

File: packages/coinstac-ui/app/render/components/field-password.jsx (hunk 16 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/field-password.jsx:23]: I guess React doesn't like string refs anymore: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-string-refs.md

File: packages/coinstac-ui/app/render/components/projects/form-project.jsx (hunk 17 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/projects/form-project.jsx]: Several were commented out before I got in there, so I didn't want to start deleting anything without someone else taking a look first. I've deleted them now along with the `//TODO`s listed in there. Let me know if I need to put them back.

File: packages/coinstac-ui/app/render/index.js (hunk 19 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/index.js]: Good point. I was looking for a way to get around the `PropTypes.object` error early on, before you advised me to cancel that rule, and put this in place before immediately forgetting I had. I've reverted back to the way it was before.

File: packages/coinstac-ui/app/render/state/ducks/util.js (hunk 20 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/state/ducks/util.js:3]: The linter wants `default` in there for cases when it's the only export: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/prefer-default-export.md

I could write it differently, but this way retained the function title, which I thought might be good for self-documenting purposes.

File: packages/coinstac-ui/app/render/components/dashboard-consortia.jsx (hunk 14 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/dashboard-consortia.jsx:6]: 😣

File: packages/coinstac-ui/app/render/components/consortium/consortium-result.jsx (hunk 13 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/consortium/consortium-result.jsx]: 💥

File: packages/coinstac-ui/app/render/components/consortium/consortium-form.jsx (hunk 12 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/consortium/consortium-form.jsx]: 💥

File: packages/coinstac-ui/app/render/components/consortium/consortium-computation-selector.jsx (hunk 11 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/consortium/consortium-computation-selector.jsx]: :boom:

File: packages/coinstac-ui/app/render/components/projects/project-covariates-mapper.jsx (hunk 18 above)
Comment (tkah) [reply packages/coinstac-ui/app/render/components/projects/project-covariates-mapper.jsx]: Eh, some syntax confusion on my part. Got it sorted on this next commit.

File: packages/coinstac-client-core/package.json (hunk 2 above)
Comment (swashcap) [reply packages/coinstac-client-core/package.json:44]: Indeed. No idea what that actually does, it’s not required anywhere (57c579675033d1eb3dcd12fd0d790a6f1a87503c).

File: packages/coinstac-ui/app/render/components/dashboard.jsx (hunk 15 above)
Comment (swashcap) [reply packages/coinstac-ui/app/render/components/dashboard.jsx:39]: :+1: Makes sense.

File: packages/coinstac-ui/app/render/components/field-password.jsx (hunk 16 above)
Comment (swashcap) [reply packages/coinstac-ui/app/render/components/field-password.jsx:23]: Looks like the React team updated the docs, too. Seems like this would have negative performance implications, but ¯\\\_(ツ)_/¯

File: packages/coinstac-ui/app/render/state/ducks/util.js (hunk 20 above)
Comment (swashcap) [reply packages/coinstac-ui/app/render/state/ducks/util.js:3]: Ah, makes sense.

File: packages/coinstac-client-core/src/model-service.js (hunk 4 above)
Comment (swashcap) [reply packages/coinstac-client-core/src/model-service.js]: Let’s add per-method `// eslint-ignore-line`s for now. Little less confusing.

**Review by swashcap — APPROVED** [review]

**Comment by swashcap** [comment]
Aside from the `modelHooks…` stuff this is looking good.

@tkah Could you fix that and the merge conflicts? I’d say it's good-to-go otherwise.

**Comment by swashcap** [comment]
Oh, I guess the e2e tests fail:

```
  Testing::e2e
    ✓ opens a single window
    ✓ displays the correct title
    1) authenticates demo user
    2) accesses the Add Consortium page
    3) creates a consortium
    4) accesses the New Files Collection page
    5) adds files to a consortium
    6) runs a computation
    7) displays results
    8) deletes files
    9) deletes a consortium
    10) logs out


  2 passing (15s)
  10 failing

  1) Testing::e2e authenticates demo user:
     Error: element (".user-account-name") still not existing after 4000ms
      at elements(".user-account-name") - isExisting.js:46:17
      at isExisting(".user-account-name") - waitForExist.js:67:22

  2) Testing::e2e accesses the Add Consortium page:
     Error: An element could not be located on the page using the given search parameters.
      at element("a=Consortia") - click.js:12:17
      at isVisible("h1=Add New Consortium") - test.js:59:8

  3) Testing::e2e creates a consortium:
     Error: An element could not be located on the page using the given search parameters ("[name='label']").
      at elements("[name='label']") - setValue.js:29:17
      at getText("div.panel-heading h4.panel-title a") - test.js:82:8

  4) Testing::e2e accesses the New Files Collection page:
     Error: An element could not be located on the page using the given search parameters.
      at element("a=My Files") - click.js:12:17
      at isVisible("h1=New Files Collection") - test.js:90:8

  5) Testing::e2e adds files to a consortium:
     Error: An element could not be located on the page using the given search parameters ("#form-project-name").
      at elements("#form-project-name") - setValue.js:29:17
      at getText("div.panel-heading h4.panel-title a") - test.js:103:8

  6) Testing::e2e runs a computation:
     Error: An element could not be located on the page using the given search parameters.
      at element("button=Run Computation") - click.js:12:17
      at waitForExist("span=Complete", 4000) - test.js:109:8

  7) Testing::e2e displays results:
     Error: An element could not be located on the page using the given search parameters.
      at element("a=View Results") - click.js:12:17
      at waitForExist("span=Complete!", 4000) - test.js:116:8

  8) Testing::e2e deletes files:
     Error: An element could not be located on the page using the given search parameters.
      at element("a=My Files") - click.js:12:17
      at isVisible("div.panel-heading") - test.js:124:8

  9) Testing::e2e deletes a consortium:
     Error: An element could not be located on the page using the given search parameters.
      at element("a=Consortia") - click.js:12:17
      at isVisible("div.panel-heading") - test.js:132:8

  10) Testing::e2e logs out:
     Error: An element could not be located on the page using the given search parameters.
      at element("button=Log Out") - click.js:12:17
      at waitForExist("button=Log In", 4000) - test.js:138:8
```

😕
